### ISSUE TITLE
(ERR): hisat2-align exited with value 141

### ISSUE THREAD

**Issue by tAndreani** [FelixKrueger/Bismark#353]
Hello,

i am running my pipeline for bisulfite WGBS data using WDL as a workflow language.

When it is time to map the reads the workflow stops with an error that is due to hisat2. You can find the error at the end of the file [here](https://github.com/FelixKrueger/Bismark/files/4808469/error_hisat2.txt). What could be the reason? As memory (that could be a potential reason for the problem), I am using 200 GB so I do not think is a memory problem.

What else could be? Should I write in hisat2 GitHub instead?

Thanks in advance.

**Comment by FelixKrueger** [comment]
Hi Tommaso,

I am not exactly sure what the reason is here to be honest. If you have access to your cluster without using the workflow, could you try to run the Bismark command its own, like so:

```
bismark --genome /genome/folder/ --hisat2 -1 SRR5395020.CGAATA.R1.trimmed.gz -2 SRR5395020.CGAATA.R1.trimmed.gz
```

This way you could monitor how much memory the process is actually trying to consume. A few. additional comments:
- which genome are you trying to use? Some genome versions contain a lot of scaffolds and patch regions, that make the genome huge
- Using --non_directional will need 4 threads of HISAT2, therefore this will roughly be 4x the size of the genome (which needs to be held in memory once in addition to that as well). Are you really sure that this is needed? I am getting some mixed signals here, I just had a look at the GEO page, and it seems that they used the Accel Swift kit. This kit typically generates directional data (but requires extensive trimming of R2 from the 5' end, e.g. using a command like: `trim_galore --clip_r1 10 --clip_r2 20 --paired SRR5395020.CGAATA.R1.fastq.gz SRR5395020.CGAATA.R2.fastq.gz`; the data section on GEO appears to have used the option `--pbat`, and finally, you seem to be using `--non_directional`. Most likely, only one of these three options is the right one to use. I'm happy to take another look at this if you could share the FastQC profile (just zip the .html file).
- I think paired-end alignments also need a bit more memory than single-end alignments
- the `-p 4` might also take additional memory. For a first test I would just drop this option
- if you are not sequencing (bisulfite converted) RNA-seq data, you should probably not be using HISAT2 in its default mode, as it will attempt to find splice junctions. If the data is standard BS-seq, then you should use `--no_splicing` (please check what the option is called) to prevent this behaviour
- to see if HISAT2 is causing the issue, just drop `--hisat2` and see if Bowtie2 does the job. You might have to install it obviously, or add it to your docker image first.

Once you have figured out whether Bismark runs and completes on its own (which might give also you an idea about the use of resources) you should be in a better position to say whether it is the Bismark itself, or your workflow that is causing the issues at hand.

Cheers, Felix

**Comment by tAndreani** [comment]
Hello Felix,  

thanks for the quick reply. I am using these data for a year and the option `--non-directional` is required to have a good alignment. I have this same pipeline (and bismark using `--hisat2` mapper) that run perfectly in clusters with SLURM and LSF scheduler but this error happens when i run the pipeline in the cloud.

I now have tried `--pbat `(expecting that did not work) as also the `--no-spliced-alignment` option and in both case i had the same error as before that you can see [here](https://github.com/FelixKrueger/Bismark/files/4808533/error_hisat2_second.txt). 

I do not know I will try with bowtie2 to see if i have the same problem.

Thanks for the suggestions.

**Comment by daler** [comment]
In case anyone else runs across this . . .

I was getting the exit code 141 problem in a recently-built conda environment . . . but did not if I switched to an older conda environment.

After doing the diffs between those environments and building many test conda environments, I finally narrowed down the problem to the conda solver. For some reason, installing _newer_ bismark and bowtie results in an _older_ samtools.

TL;DR: if you are using conda to build your environment, try pinning `samtools>=1.10` to fix the error. This solved it for me.

**Comment by FelixKrueger** [comment]
> In case anyone else runs across this . . .
> 
> I was getting the exit code 141 problem in a recently-built conda environment . . . but did not if I switched to an older conda environment.
> 
> After doing the diffs between those environments and building many test conda environments, I finally narrowed down the problem to the conda solver. For some reason, installing _newer_ bismark and bowtie results in an _older_ samtools.
> 
> TL;DR: if you are using conda to build your environment, try pinning `samtools>=1.10` to fix the error. This solved it for me.

Thanks for posting this in case someone finds themselves in a similar pickle in the future, I am glad it is now working!

**Comment by ikkextoch** [comment]
I had this issue when using head on the output of hisat2 with a reference index that was build with a `fasta.gz` file containing the gzipped reference sequence of my species of interest. Building the index with an uncompressed fasta solved the mapping issue.
